src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java

```diff
@@ -345,40 +346,45 @@ public static synchronized boolean addEventToJenkinsJobRun(Build build)
         return false;
     }
 
-    public static synchronized void addBuildToNoBCList(Build build) {
+    private static void addBuildToNoBCList(Build build) {
         // should have been caught upstack, but just in case since public method
         if (!OpenShiftUtils.isPipelineStrategyBuild(build))
             return;
-        buildsWithNoBCList.add(build);
-    }
-
-    private static synchronized void removeBuildFromNoBCList(Build build) {
-        buildsWithNoBCList.remove(build);
+        try {
+          buildsWithNoBCList.add(build);
+        } catch (ConcurrentModificationException | IllegalArgumentException |
+          UnsupportedOperationException | NullPointerException e) {
+          logger.log(Level.WARNING,"Failed to add item " +
+            build.getMetadata().getName(), e);
+        }
     }
 
-    private static synchronized void clearNoBCList() {
-        buildsWithNoBCList.clear();
+    private static void removeBuildFromNoBCList(Build build) {
+          buildsWithNoBCList.remove(build);
     }
 
     // trigger any builds whose watch events arrived before the
     // corresponding build config watch events
-    public static synchronized void flushBuildsWithNoBCList() {
-        HashSet<Build> clone = (HashSet<Build>) buildsWithNoBCList.clone();
-        clearNoBCList();
-        for (Build build : clone) {
-            WorkflowJob job = getJobFromBuild(build);
-            if (job != null)
-                try {
-                    logger.info("triggering job run for previously skipped build "
-                            + build.getMetadata().getName());
-                    triggerJob(job, build);
-                } catch (IOException e) {
-                    logger.log(Level.WARNING, "flushCachedBuilds", e);
-                }
-            else
-                addBuildToNoBCList(build);
+    public static void flushBuildsWithNoBCList() {
+        for (Build build : buildsWithNoBCList) {
+          WorkflowJob job = getJobFromBuild(build);
+          if (job != null) {
+            try {
+              logger.info("triggering job run for previously skipped build "
+                + build.getMetadata().getName());
+              triggerJob(job, build);
+            } catch (IOException e) {
+              logger.log(Level.WARNING, "flushCachedBuilds", e);
+            }
+            try {
+              removeBuildFromNoBCList(build);
+            } catch (UnsupportedOperationException | IllegalStateException |
+              NullPointerException | ClassCastException e) {
+              logger.log(Level.WARNING, "Failed to remove item" +
+                build.getMetadata().getName(), e);
```

Comment:
we have to ensure that we do not leave any stale entries if a remove failed.

So please add this:
1) add a boolean, called anyRemoveFailures, initialized to false
2) if an exception on the remove occurs, set anyRemoveFailures to true

Outside the for loop:

```
if (anyRemoveFailures && buildsWithNoBCList.size() > 0) {
   buildsWithNoBCList.clear()
}
```

If no exceptions occurred, the compiler will bail and `buildsWithNoBCList.size()` will not even be called